File: python/SelfParentImputer.py
from __future__ import annotations

# coding: utf-8
# SelfParentImputer.py
# 親が自殖で後代がimputedなとき親をimputeする

from math import log
from typing import List, Tuple

from GenoRecord import GenoRecord
from VCFHMM import *
from Genotype import Genotype


#################### SelfParentImputer ####################

MIN_PROB = -1e300

class SelfParentImputer(VCFHMM[GenoRecord]):
	DP = List[Tuple[float, int]]	# (log of probability, prev h)
	
	def __init__(self, records: list[GenoRecord],
							ref_haps: list[list[int]],
							iprog: int, map_: Map, w: float) -> None:
		VCFHMM.__init__(self, records, map_)
		self.records: list[GenoRecord] = records
		self.ref_haps = ref_haps
		self.ic = iprog		# 何番目の後代か
		self.prev_h_table = self.collect_possible_previous_hidden_states()
		self.Epc = self.calc_Epc(w)
		# crossover values
		# 後代
		self.Cc = [ Map.Kosambi(self.dist(r1, r2))
							for r1, r2 in zip(self.records, self.records[1:]) ]
		K = 5		# 親の遷移確率の子に対する倍率
		# 親は直接の後代でないので、乗り換え確率を高くする
		self.Cp = [ Map.Kosambi(self.dist(r1, r2) * K)
							for r1, r2 in zip(self.records, self.records[1:]) ]
	
	# 親と後代のnon-phased genotypeからの後代の排出確率
	def calc_Epc(self, w: float) -> list[list[float]]:
		E1: list[list[float]] = [
				[ 1.0-w*2,	w/2,	   w/2,    w ],
				[ 0.25-w/4, 0.5-w/2, 0.25-w/4, w ],
				[ 0.25-w/4, 0.5-w/2, 0.25-w/4, w ],
				[ w/2,		w/2,	 1.0-w*2,  w ]
		]
		E = [ [ log(p) for p in v ] for v in E1 ]
		return E
	
	def NH(self) -> int:
		return len(self.ref_haps)
	
	def num_progenies(self) -> int:
		return len(self.records[0].geno) - 9
	
	def decode_state(self, h: int) -> tuple[int, int, int, int]:
		# 1bit目: 後代の左は親のどちらのハプロタイプ由来か
		# 2bit目: 後代の右は親のどちらのハプロタイプ由来か
		# 3bit以降: 親のハプロタイプの由来
		# 親のハプロタイプは少なくとも一方は後代と同じでないといけない
		hp, hc = divmod(h, 4)
		if hp < self.NH() * 2 + 4:
			h2, h1 = divmod(hp, self.NH() + 2)
		else:
			h2, h1 = divmod(hp - self.NH() * 2, 2)
		return (h1, h2, hc & 1, hc >> 1)
	
	def num_states(self) -> int:
		return (self.NH() + 1) * 16
	
	def allele(self, h: int, i: int, prog_gt: int) -> int:
		# h: 後代1, 後代2, リファレンスハプロタイプ
		if h < 2:
			return (prog_gt >> h) & 1
		else:
			return self.ref_haps[h-2][i]
	
	def parent_genotype(self, h: int, i: int, prog_gt: int) -> int:
		hp1, hp2, hc1, hc2 = self.decode_state(h)
		a1 = self.allele(hp1, i, prog_gt)
		a2 = self.allele(hp2, i, prog_gt)
		return a1 | (a2 << 1)
	
	def progeny_genotype(self, h: int, i: int, prog_gt: int) -> int:
		hp1, hp2, hc1, hc2 = self.decode_state(h)
		gt_parent = self.parent_genotype(h, i, prog_gt)
		return ((gt_parent >> hc1) & 1) | ((gt_parent >> hc2) & 1)
	
	# phasedなGenotypeに対してphasedなGenotypeが排出される確率
	def phased_emission_probability(self, h: int, i: int,
												prog_gt: int) -> float:
		phased_prog_gt = self.progeny_genotype(h, i, prog_gt)
		d = phased_prog_gt ^ prog_gt
		return log((0.99 if (d & 1) == 0 else 0.01) *
					(0.99 if (d >> 1) == 0 else 0.01))
	
	def non_phased_emission_probability(self, ocs: list[int],
												parent_gt: int) -> float:
		return sum(self.Epc[parent_gt][oc] for oc in ocs)
	
	# i: record index
	def emission_probability(self, h: int, i: int, op: int,
									prog_gt: int, ocs: list[int]) -> float:
		# imputedな後代も排出確率を計算する
		parent_gt = self.parent_genotype(h, i, prog_gt)
		Ep = self.E[parent_gt][op]
		Ec = self.phased_emission_probability(h, i, prog_gt)
		Ecs = self.non_phased_emission_probability(ocs, parent_gt)
		return Ep + Ec + Ecs
	
	def initialize_dp(self) -> list[DP]:
		M = len(self.records)	# マーカー数
		L = self.num_states()
		dp = [ [ (MIN_PROB, 0) ] * L for _ in range(M) ]
		record = self.records[0]
		op = record.unphased(0)		# observed parent
		prog_gt = record.geno[self.ic+1] & 3	# phased progeny
		ocs = [ record.unphased(j+1) for j in range(self.num_progenies())
														if j != self.ic ]
		for h in range(L):		# hidden state
			E_all = self.emission_probability(h, 0, op, prog_gt, ocs)
			dp[0][h] = (E_all, h)
		return dp
	
	# hidden stateに対して、可能な前のhidden stateを集めておく
	def collect_possible_previous_hidden_states(self) -> list[list[int]]:
		# 親の両側ともリファレンスハプロタイプということはない
		L = self.num_states()
		prev_h_table: list[list[int]] = [ [] for _ in range(L) ]
		for h in range(L):		# hidden state
			hp1, hp2, hc1, hc2 = self.decode_state(h)
			# 両側乗り換えることはないとする
			# non-phasedの親のあり得る前の状態
			for prev_h in range(L):
				prev_hp1, prev_hp2, prev_hc1, prev_hc2 = \
												self.decode_state(prev_h)
				diff_counter = ((1 if hp1 != prev_hp1 else 0) +
								(1 if hp2 != prev_hp2 else 0) +
								(1 if hc1 != prev_hc1 else 0) +
								(1 if hc2 != prev_hc2 else 0))
				if diff_counter <= 1:
					prev_h_table[h].append(prev_h)
		
		return prev_h_table
	
	def transition_probability(self, h: int, prev_h: int,
										cc: float, cp: float) -> float:
		hp1, hp2, hc1, hc2 = self.decode_state(h)
		prev_hp1, prev_hp2, prev_hc1, prev_hc2 = self.decode_state(prev_h)
		# 遷移確率 0なら(1-c)、1ならcを掛ける
		return (log(cp if hp1 != prev_hp1 else 1.0 - cp) +
				log(cp if hp2 != prev_hp2 else 1.0 - cp) +
				log(cc if hc1 != prev_hc1 else 1.0 - cc) +
				log(cc if hc2 != prev_hc2 else 1.0 - cc))
	
	def update_dp(self, i: int, dp: list[DP]) -> None:
		L = self.num_states()
		record = self.records[i]
		cc = self.Cc[i-1]	# 遷移確率
		cp = self.Cp[i-1]	# 親の遷移確率
		op = record.unphased(0)		# observed parent
		prog_gt = record.geno[self.ic+1] & 3	# phased progeny
		ocs = [ record.unphased(j+1) for j in range(self.num_progenies())
															if j != self.ic ]
		
		for h in range(L):		# hidden state
			E_all = self.emission_probability(h, i, op, prog_gt, ocs)
			
			for prev_h in self.prev_h_table[h]:
				T_all = self.transition_probability(h, prev_h, cc, cp)
				
				prob = dp[i-1][prev_h][0] + (T_all + E_all)
				dp[i][h] = max(dp[i][h], (prob, prev_h))
	
	def update_genotypes(self, hs: list[int]) -> None:
		M = len(self.records)
		for i in range(M):
			record = self.records[i]
			prog_gt = record.geno[self.ic+1] & 3
			parent_gt = self.parent_genotype(hs[i], i, prog_gt)
			record.geno[0] = parent_gt | 4
	
	def impute(self) -> None:
		# DP
		# 外側から、マーカー、ハプロタイプの状態
		M = len(self.records)
		dp = self.initialize_dp()
		for i in range(1, M):
			self.update_dp(i, dp)
		
		hs = self.trace_back(dp)
		self.update_genotypes(hs)

File: src/SelfParentImputer.cpp
#include <cmath>
#include "../include/SelfParentImputer.h"
#include "../include/common.h"

using namespace std;

SelfParentImputer::SelfParentImputer(const vector<GenoRecord *>& rs,
									 const vector<vector<int>>& ref_hs,
									 size_t iprog, const Map& map_, double w) :
			VCFHMM(rs, map_, w), ref_records(rs), ref_haps(ref_hs),
			ic(iprog), prev_h_table(collect_possible_previous_hidden_states()),
			Cc(calc_Cc(rs)), Cp(calc_Cp(rs)),
			Epc{{log(1.0-w*2),  log(w/2),     log(w/2),      log(w)},
				{log(0.25-w/4), log(0.5-w/2), log(0.25-w/4), log(w)},
				{log(0.25-w/4), log(0.5-w/2), log(0.25-w/4), log(w)},
				{log(w/2),      log(w/2),     log(1.0-w*2),  log(w)}}
			{ }

vector<double> SelfParentImputer::calc_Cc(
							const vector<GenoRecord *>& rs) const {
	const size_t	M = rs.size();
	vector<double>	Cc(M-1);
	for(size_t i = 0; i < M - 1; ++i) {
		Cc[i] = Map::Kosambi(dist(rs[i], rs[i+1]));
	}
	return Cc;
}

vector<double> SelfParentImputer::calc_Cp(
							const vector<GenoRecord *>& rs) const {
	const double	K = 5.0;
	const size_t	M = rs.size();
	vector<double>	Cp(M-1);
	for(size_t i = 0; i < M - 1; ++i) {
		Cp[i] = Map::Kosambi(dist(rs[i], rs[i+1]) * K);
	}
	return Cp;
}

tuple<int, int, int, int> SelfParentImputer::decode_state(int h) const {
	// 1st bit: Which parent's haplotype does the left side of the progeny come from?
	// 2nd bit: Which parent's haplotype does the right side of the progeny come from?
	// At least one of the parent's haplotypes must be the same as the progeny's
	const int	hc = h & 3;
	const int	hp = h >> 2;
	if(hp < (int)NH() * 2 + 4) {
		const int	hp1 = hp % (NH() + 2);
		const int	hp2 = hp / (NH() + 2);
		return make_tuple(hp1, hp2, hc & 1, hc >> 1);
	}
	else {
		const int	hp1 = (hp - NH() * 2) & 1;
		const int	hp2 = (hp - NH() * 2) >> 1;
		return make_tuple(hp1, hp2, hc & 1, hc >> 1);
	}
}

double SelfParentImputer::phased_emission_probability(int h, size_t i,
															int prog_gt) const {
	const int	phased_prog_gt = progeny_genotype(h, i, prog_gt);
	const int	d = phased_prog_gt ^ prog_gt;
	return log((((d & 1) == 0) ? 0.99 : 0.01) *
				(((d >> 1) == 0) ? 0.99 : 0.01));
}

double SelfParentImputer::non_phased_emission_probability(
												const vector<int>& ocs,
												int parent_gt) const {
	double	s = 0.0;
	for(auto p = ocs.begin(); p != ocs.end(); ++p) {
		s += Epc[parent_gt][*p];
	}
	return s;
}

double SelfParentImputer::emission_probability(int h, size_t i, int op,
												int prog_gt,
												const vector<int>& ocs) const {
	const int	parent_gt = parent_genotype(h, i, prog_gt);
	const double	Ep = E[parent_gt][op];
	const double	Ec = phased_emission_probability(h, i, prog_gt);
	const double	Ecs = non_phased_emission_probability(ocs, parent_gt);
	return Ep + Ec + Ecs;
}

double SelfParentImputer::transition_probability(size_t i,
													int prev_h, int h) const {
	const double	cp = Cp[i-1];
	const double	cc = Cc[i-1];
	const auto	t1 = decode_state(h);
	const auto	t2 = decode_state(prev_h);
	return (log(get<0>(t1) != get<0>(t2) ? cp : 1.0 - cp) +
			log(get<1>(t1) != get<1>(t2) ? cp : 1.0 - cp) +
			log(get<2>(t1) != get<2>(t2) ? cc : 1.0 - cc) +
			log(get<3>(t1) != get<3>(t2) ? cc : 1.0 - cc));
}

vector<SelfParentImputer::DP> SelfParentImputer::initialize_dp() const {
	const size_t	L = num_states();
	vector<DP>	dp(M(), DP(L, pair<double, int>(MIN_PROB, 0)));
	const GenoRecord	*record = records[0];
	const int	op = record->unphased(0);
	const int	prog_gt = record->get_geno()[ic+1] & 3;
	vector<int>	ocs;
	for(size_t j = 0; j < num_progenies(); ++j) {
		if(j != ic) {
			ocs.push_back(record->unphased(j+1));
		}
	}
	for(int h = 0; h < (int)L; ++h) {
		const double	E_all = emission_probability(h, 0, op, prog_gt, ocs);
		dp[0][h] = make_pair(E_all, h);
	}
	return dp;
}

// Collect possible previous hidden states for the hidden state.
vector<vector<int>>
SelfParentImputer::collect_possible_previous_hidden_states() const {
	const size_t	L = num_states();
	vector<vector<int>>	prev_h_table(L, vector<int>());
	for(int h = 0; h < (int)L; ++h) {	// hidden state
		const auto	t1 = decode_state(h);
		for(int prev_h = 0; prev_h < (int)L; ++prev_h) {
			const auto	t2 = decode_state(prev_h);
			int	diff_counter = (get<0>(t1) != get<0>(t2) ? 1 : 0) +
							   (get<1>(t1) != get<1>(t2) ? 1 : 0) +
							   (get<2>(t1) != get<2>(t2) ? 1 : 0) +
							   (get<3>(t1) != get<3>(t2) ? 1 : 0);
			if(diff_counter <= 1) {
				prev_h_table[h].push_back(prev_h);
			}
		}
	}
	
	return prev_h_table;
}

void SelfParentImputer::update_dp(size_t i, vector<DP>& dp) const {
	const size_t	L = num_states();
	const GenoRecord	*record = records[i];
	const int	op = record->unphased(0);
	const int	prog_gt = record->get_geno()[ic+1] & 3;
	vector<int>	ocs;
	for(size_t j = 0; j < num_progenies(); ++j) {
		if(j != ic) {
			ocs.push_back(record->unphased(j+1));
		}
	}
	for(int h = 0; h < (int)L; ++h) {
		const double	E_all = emission_probability(h, i, op, prog_gt, ocs);
		
		const auto&	prev_hs = prev_h_table[h];
		for(auto p = prev_hs.begin(); p != prev_hs.end(); ++p) {
			const int	prev_h = *p;
			const double	Tp = transition_probability(i, prev_h, h);
			const double	prob = dp[i-1][prev_h].first + (Tp + E_all);
			dp[i][h] = max(dp[i][h], make_pair(prob, prev_h));
		}
	}
}

void SelfParentImputer::update_genotypes(const vector<int>& hs) {
	for(size_t i = 0; i < this->M(); ++i) {
		GenoRecord	*record = records[i];
		const int	prog_gt = record->get_geno()[ic+1];
		const int	phased_gt = parent_genotype(hs[i], i, prog_gt);
		record->set_geno(0, phased_gt | 4);
	}
}

void SelfParentImputer::impute() {
	// DP
	vector<DP>	dp = initialize_dp();
	for(int i = 1; i < (int)this->M(); ++i) {
		update_dp(i, dp);
	}
	
	const vector<int>	hs = trace_back(dp);
	update_genotypes(hs);
}
